src/query_plan/generate.rs

```diff
@@ -0,0 +1,321 @@
+use crate::query_plan::QueryPlanCost;
+use std::collections::VecDeque;
+
+type Choices<T> = Vec<Option<T>>;
+
+struct Partial<Plan, Element> {
+    partial_plan: Plan,
+    partial_cost: Option<QueryPlanCost>,
+    remaining: std::vec::IntoIter<Choices<Element>>,
+    is_root: bool,
+    index: Option<usize>,
+}
+
+struct Extracted<Element> {
+    extracted: Element,
+    is_last: bool,
+}
+
+/// Given some initial partial plan and a list of options for the remaining parts
+/// that need to be added to that plan to make it complete,
+/// this method "efficiently" generates (or at least evaluate) all the possible complete plans
+/// and the returns the "best" one (the one with the lowest cost).
+///
+/// Note that this method abstracts the actual types of both plans
+/// and additional elements to add to the plan,
+/// and this both for clarity and to make testing of this method easier.
+/// But type parameter `Plan` should be though of as abstracting a query plan
+/// (in practice, it is instanciated to a pair of a (`DependencyGraph`, corresponding `PathTree`)),
+/// whith `Eelement` should be though of as an additional element
+/// to add to the plan to make it complete
+/// (instanciated in practice by a `PathTree` for ... reasons ...
+/// but one that really correspond to a single `GraphPath`).
```

Comment:
fixing some typos

```suggestion
/// Note that this method abstracts the actual types of both plans (type parameter `Plan`) and 
/// additional elements (type parameter `Element`) to add to the plan. This is done for both the 
/// clarity and to make testing of this method easier. 
///
/// Type parameter `Plan` should be thought of as abstracting a query plan but in practice, it is 
/// instantiated to a pair of a (`DependencyGraph`, corresponding `PathTree`).
/// 
/// Type parameter `Element` should be thought of as an additional element to add to the plan to 
/// make it complete. It instantiated in practice by a `PathTree` (for ... reasons ...) but one 
/// that really correspond to a single `GraphPath`.
```

Reply:
I reformatted your edits to preserve [semantic line breaks](https://sembr.org/) and make the diff easier to read.